### PR TITLE
Add History Archival feature

### PR DESCRIPTION
**Pull request by davelopez**
Implementation of #14771

Depends on #16075 and #16143

## UI
There is a new option on the histories menu to let you "archive" a history (moving it out of your *active histories*)

image

When `Archiving` a history, depending on the server configuration, we are presented with 2 options, one to *Keep the storage space* and another to *free the storage space* taken by the history.

### Keep storage space (option A)
image

This option will just mark the history as being *Archived* and will no longer be shown in the regular histories listing or the "History Selector". 
Some characteristics of this kind of archived histories:
- The contents of the history will remain as they are.
- The UI will not allow certain operations on the history (the API will still allow them until a *Read Only* mode for histories is implemented on the backend as a future feature)
- The history can be `restored` at any moment, effectively moving it back to your active histories.

### Free storage space (option B)
image

This option will allow the user to package & export the history to a **permanent remote source** as a backup snapshot and then purge the history and its contents to free up storage space.

After generating the *export record* the user should acknowledge that the original history will be permanently deleted and then proceed to archive it.

image

Some characteristics of this kind of archived histories:
- If there is already a recent (up-to-date) export record to a **permanent file source** you don't need to generate another one.
- These histories are purged. So they are read-only by definition.
- You can always **import a new copy** out of the associated *export record* in case you want to restore it, this will import back the backup snapshot that you associated when archiving it.

### Archived Histories
Just a simple list where you can explore your archived histories and restore or reimport them as necessary.

image

---

## Backend

The backend implementation builds on top of #14839, this allows to optionally associate an export record at the time of archival so the actual contents of the history can be purged.

Only export records targeting a **permanent file source** are considered for archival backup.

### Changes in the database
- The `history` table now contains 2 new columns:
  - `archived`: just a flag to indicate that this history is no longer part of the active histories of the user.
  - `archive_export_id`: the optional ID of an export record that contains a persisted backup snapshot of the history.

### New API routes
- `GET /api/histories/archived`: returns the list of archived histories of the user. Supports filtering and pagination as other similar endpoints.
- `POST /api/histories/{history_id}/archive`: marks a history as archived. Optionally it can associate an export record and purge the history.
- `PUT /api/histories/{history_id}/archive/restore`: will restore the history (unarchive, move it back to your active histories). Purged histories are not restored by default but you can still "force" the unarchival by using the `force=true` query parameter. It will be then like one of your regular purged histories.

---

## How to test the changes?
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

### CODE REVIEW

**Comment by jdavcs** [comment]
@davelopez so here's what's happening with the migrations error.  There are two. The first one (the one we discussed) happens because of the `index` parameter is ignored ([ref: see gray box](https://alembic.sqlalchemy.org/en/latest/ops.html#alembic.operations.Operations.add_column)). A separate statement is needed. 

Once that one is fixed, another error is raised, that one is SQLite-specific: handling foreign keys in batch operations mode requires separate statements ([here's a detailed discussion](https://alembic.sqlalchemy.org/en/latest/batch.html#dealing-with-constraints)).

So, my suggestions are as follows:
1. I think separating this into 2 revisions might be a little cleaner, given the constraints? (that's how I tested, but having everything in 1 file should work too)
2. For the `archived` column:
```py
def upgrade():
    with transaction():
        add_column(
            table_name,
            sa.Column(column_name, sa.Boolean(), default=False, server_default=sa.false()),
        )
        create_index(index_name, table_name, [column_name])

def downgrade():
    with transaction():
        drop_index(index_name, table_name)
        drop_column(table_name, column_name)
```
3. For the `archive_export_id` column:
```py
fk_name = f"fk_{table_name}_{column_name}"

column = sa.Column(
    column_name,
    sa.Integer,
    nullable=True,
    default=None,
)

def upgrade():
    with transaction():
        add_column(table_name, column)
        create_foreign_key(fk_name, table_name, ref_table_name, [column_name], ["id"])

def downgrade():
    with transaction():
        drop_constraint(fk_name, table_name)
        drop_column(table_name, column_name)
```
(We definitely need an automated naming convention - to ensure that the indexes and constraints automatically created from our model definitions match those created via migrations. It's on my todo list.)

**Comment by davelopez** [comment]
Thank you again for all the help @jdavcs!

> I think separating this into 2 revisions might be a little cleaner, given the constraints? (that's how I tested, but having everything in 1 file should work too)

I decided to keep both in the same revision file, but I'm happy to split them if this is important. The updated commit is [2effa4a](https://github.com/galaxyproject/galaxy/pull/16003/commits/2effa4a4a0dc71974aa0915c62f58a919534e4d3).

**Comment by davelopez** [comment]
I made changes to address the comments in the UI/UX meeting and the issues about histories not being immutable after purging or archiving should be solved or dramatically reduced by #16143

The only thing that remains is the issue with the [fetcher and the PUT with query params](https://github.com/galaxyproject/galaxy/pull/16003/files#diff-a1ff7e7ceba68255d276ca503a7f8a3e5052f96f805fa190a833285600fca937R87) getting ignored, but I can change it to use Axios temporarily until we update the library if it is an upstream bug.

So this is ready for a closer look at the code from whoever wants to take a look, thanks! :pray:

**Comment by mvdbeek** [comment]
> Replaced with axios as a workaround for now.

Can you try `.create({force: undefined})` ?

**Comment by davelopez** [comment]
> Can you try .create({force: undefined}) ?

Thank you Marius, that did the trick! the only little issue now is that TS doesn't like that apparently `create` accepts 0 arguments in its type definition... but still better than the other workaround! :+1: do maybe we can relax eslint to allow "ts-ignore with comments"? or is there an updated version of the library that has the correct typings?

**Comment by mvdbeek** [comment]
I think so, can you try `@qdrant/openapi-typescript-fetch` ?

**Comment by davelopez** [comment]
> can you try @qdrant/openapi-typescript-fetch ?

I tried but it has the same issue with `create` taking 0 arguments :(

I will turn the eslint rule to warning for now as it is allowed inside .vue files anyway, and we can remove it when swapping to the new way in openapi-typescript down the road.

File: client/src/components/History/Archiving/HistoryArchiveExportSelector.vue [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+        emit("onArchive", mostUpToDateExport.value.id);
+    }
+}
+</script>
+
+<template>
+    <div>
+        <p>
+            If you are not planning to use this history in the near future, you can
+            <b>archive and delete its contents to free up valuable disk space</b>.
```
Comment (afgane) [inline client/src/components/History/Archiving/HistoryArchiveExportSelector.vue]:
```suggestion
            <b>archive and delete its contents to free up disk space</b>.
```

File: client/src/components/History/Archiving/HistoryArchiveExportSelector.vue [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+        </p>
+        <p>
+            In order to be able to recover your archived history later, you need to export it first to a permanent
+            remote location. Then you will be able to import it back to Galaxy from the remote source (as a new copy).
```
Comment (afgane) [inline client/src/components/History/Archiving/HistoryArchiveExportSelector.vue]:
```suggestion
            To be able to recreate your archived history later, you need to export it first to a permanent
            remote location. Then you will be able to import it back to Galaxy from the remote source, as a new copy.
```

File: client/src/components/History/Archiving/HistoryArchiveExportSelector.vue [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+            </p>
+            <p>{{ exportErrorMessage }}</p>
+        </b-alert>
+        <div v-else-if="mostUpToDateExport && mostUpToDateExportIsReady">
+            <b-alert show variant="success">
+                <p>
+                    <b>There is an up-to-date export record of this history ready.</b>
+                </p>
+                <p>
+                    This export record will be associated with your archived history so you can recover it later by
```
Comment (afgane) [inline client/src/components/History/Archiving/HistoryArchiveExportSelector.vue]:
```suggestion
                    This export record will be associated with your archived history so you can recreate it later by
```

File: client/src/components/History/Archiving/HistoryArchiveExportSelector.vue [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+        </div>
+        <div v-else>
+            <b-alert v-if="isExportingRecord" id="generating-export-record-alert" show variant="info">
+                <loading-span message="Generating export record. This may take a while..." />
+            </b-alert>
+            <b-alert v-else show variant="info">
+                <p>
+                    <b>
+                        There is no up-to-date export record of this history. You need to create a new export record to
+                        be able to recover the history later.
```
Comment (afgane) [inline client/src/components/History/Archiving/HistoryArchiveExportSelector.vue]:
```suggestion
                        be able to recreate the history later.
```

File: client/src/components/History/Archiving/HistoryArchiveExportSelector.vue [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+                    variant="primary"
+                    @click="onCreateExportRecord">
+                    Create export record
+                </b-button>
+            </b-alert>
+        </div>
+        <p v-if="!isDeleteContentsConfirmed" class="mt-3 mb-0">
+            To continue, you need to confirm that you want to delete the contents of the original history before you can
+            archive it using the checkbox below. No worries, you will be able to recover the history later by importing
+            it from the export record above.
```
Comment (afgane) [inline client/src/components/History/Archiving/HistoryArchiveExportSelector.vue]:
```suggestion
            To continue, you need to confirm that you want to delete the contents of the original history before you can
            archive it using the checkbox below. If you created an export record above, you will be able to recreate the history later by importing
            it from the export record.
```

**Review by afgane — COMMENTED** [review]
Very cool capability - I just have a few suggestions on the wording so see what you think.

**Comment by davelopez** [comment]
Thank you @afgane! the rewording looks way better, I really appreciate it as I always struggle with that...

**Comment by davelopez** [comment]
I've fixed all the conflicts and updated de db migration script commit with the latest revision. This should be ready to go if someone has time to review it :bowing_man:

**Pull request merged**